### PR TITLE
add playwright testing and e2e testing

### PR DESCRIPTION
**Pull request by Gkrumbach07**
closes: https://github.com/opendatahub-io/odh-dashboard/issues/1043

## Description

- added e2e testing by `npm run test:e2e`
- added playwright tests for integration testing
- removed storybook play function tests
- upgraded storybook to v7
- merging testing ci into one command `npm run test`

## How Has This Been Tested?

1. run `npm run test` and check that all is passing
2. configure .env.test to point to the dashboard you wish to test for e2e
3. run `npm run test:e2e` which will run one test and cleanup if a failure occurred

## Test Impact

tests will use playwright and jest instead of storybook interaction tests

## Request review criteria:

- [x] The commits have meaningful messages (squashes happen on merge by the bot).
- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -52,5 +52,8 @@
     "jest": "^26.6.3",
     "npm-run-all": "^4.1.5",
     "prettier": "^2.2.1"
+  },
+  "devDependencies": {
```
Comment (lucferbux) [inline package.json:56]: Just a little nit, but what do you think about adding here `test:e2e` and `test:accesibility` so we have those commands exposed in the root of the repo?

Comment (Gkrumbach07) [reply package.json:56]: yeah i can add those to the top

Comment (Gkrumbach07) [reply package.json:57]: I pushed them all to the root and cleaned up unused commands.

Comment (lucferbux) [reply package.json:57]: Perfect thanks!!

**Review by lucferbux — APPROVED** [review]
/lgtm

**Comment by lucferbux** [comment]
@andrewballantyne given Juntao and alex are way focused in explainability, do you prefer me pinging you or should i ask them for review?

File: .gitignore [hunk 2]
```diff
@@ -17,4 +17,4 @@ logs
 npm-debug.log*
 yarn-debug.log*
 yarn-error.log*
-yarn.lock
+yarn.lock
```
Comment (andrewballantyne) [inline .gitignore:20]: Why this change?

File: frontend/playwright.config.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  /* Run tests in files in parallel */
+  fullyParallel: !!process.env.E2E,
+  /* Opt out of parallel tests on CI. */
+  workers: process.env.E2E ? undefined : 1,
+  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
+  reporter: 'html',
+  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
+  use: {
+    /* Base URL to use in actions like `await page.goto('/')`. */
+    baseURL: process.env.E2E ? process.env.E2E_DASHBOARD_URL : 'http://localhost:6006',
```
Comment (andrewballantyne) [inline frontend/playwright.config.ts:26]: Why this fallback url?

File: frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  await page.goto('/projects');
+  const projectTestRow = await page.locator('tr', { has: page.locator('text="e2e-test-project"') });
+  const count = await projectTestRow.count();
+  if (count > 0) {
+    await projectTestRow.getByRole('button', { name: 'Actions' }).click();
+    await page.getByRole('menuitem', { name: 'Delete project' }).click();
+    await page.getByRole('textbox', { name: 'Delete modal input' }).fill('e2e-test-project');
+    await page.getByRole('button', { name: 'Delete project' }).click();
+  }
+});
```
Comment (andrewballantyne) [inline frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts:25]: Any reason we do this in both BEFORE and AFTER? In case the test breaks midway and we want to cleanup before we start?

File: frontend/src/__tests__/integration/pages/modelServing/ModelServingGlobal.spec.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
+import { test, expect } from '@playwright/test';
+
+test('Delete model', async ({ page }) => {
+  await page.goto(
+    './iframe.html?args=&id=tests-integration-pages-modelserving-modelservingglobal--delete-model&viewMode=story',
```
Comment (andrewballantyne) [inline frontend/src/__tests__/integration/pages/modelServing/ModelServingGlobal.spec.ts:5]: What is happening here?

File: frontend/src/__tests__/integration/pages/projects/ProjectDetails.spec.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { test, expect } from '@playwright/test';
+
+test('Empty project', async ({ page }) => {
+  await page.goto(
+    './iframe.html?args=&id=tests-integration-pages-projects-projectdetails--empty-details-page&viewMode=story',
```
Comment (andrewballantyne) [inline frontend/src/__tests__/integration/pages/projects/ProjectDetails.spec.ts:5]: These items definitely need to be hidden away from our tests if this is needed...

File: package.json [hunk 7]
```diff
@@ -39,18 +39,22 @@
...
+    "test:fix-frontend": "cd ./frontend && npm run test:fix"
   },
   "dependencies": {
     "dotenv": "^8.2.0",
     "dotenv-expand": "^5.1.0",
     "jest": "^26.6.3",
     "npm-run-all": "^4.1.5",
     "prettier": "^2.2.1"
+  },
+  "devDependencies": {
```
Comment (andrewballantyne) [inline package.json:57]: Would this not be optional dependencies? Playwright is a problem for downstream installs...

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

**Review by andrewballantyne — COMMENTED** [review]
My test failed because it timed out 🤔 

image

File: frontend/package.json [hunk 8]
```diff
@@ -117,17 +117,20 @@
...
+    "@storybook/addon-interactions": "^7.0.17",
+    "@storybook/addon-links": "^7.0.17",
+    "@storybook/jest": "0.1.0",
+    "@storybook/react": "^7.0.17",
+    "@storybook/react-webpack5": "^7.0.17",
+    "@storybook/test-runner": "^0.10.0",
+    "@storybook/testing-library": "0.1.0",
     "@testing-library/dom": "^8.13.0",
     "@testing-library/jest-dom": "^5.16.5",
     "@testing-library/react": "^13.4.0",
```
Comment (andrewballantyne) [inline frontend/package.json:136]: Do we need this anymore? I think you deleted the old tests that were the last ones to rely on this, no?

File: frontend/package.json [hunk 9]
```diff
@@ -117,17 +117,20 @@
...
+    "@storybook/addon-essentials": "^7.0.17",
+    "@storybook/addon-interactions": "^7.0.17",
+    "@storybook/addon-links": "^7.0.17",
+    "@storybook/jest": "0.1.0",
+    "@storybook/react": "^7.0.17",
+    "@storybook/react-webpack5": "^7.0.17",
+    "@storybook/test-runner": "^0.10.0",
+    "@storybook/testing-library": "0.1.0",
     "@testing-library/dom": "^8.13.0",
     "@testing-library/jest-dom": "^5.16.5",
```
Comment (andrewballantyne) [inline frontend/package.json:135]: I imagine we don't need this anymore either -- we should not be testing the DOM with Jest...

File: frontend/src/__tests__/integration/pages/projects/ProjectView.spec.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+  await expect(page.getByRole('button', { name: 'Create' })).toBeEnabled();
+
+  await page.getByLabel('Resource name *').fill('%^&test-project@#$');
+  await expect(page.getByRole('button', { name: 'Create' })).toBeDisabled();
+
+  await page.getByLabel('Resource name *').fill('test.project');
+  await expect(page.getByRole('button', { name: 'Create' })).toBeDisabled();
+
+  await page.getByLabel('Resource name *').fill('test-project');
+  await expect(page.getByRole('button', { name: 'Create' })).toBeEnabled();
```
Comment (andrewballantyne) [inline frontend/src/__tests__/integration/pages/projects/ProjectView.spec.ts:38]: Am I missing something -- this does not appear to be part of the tests:
image

File: frontend/playwright.config.ts (hunk 3 above)
Comment (Gkrumbach07) [reply frontend/playwright.config.ts:26]: 6006 is the storybook port. if e2e is used then use whatever url you are testing on (by default it is localhost:4010 but it could be any dashboard url). if you are not doing e2e then it uses storybook server

File: frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts (hunk 4 above)
Comment (Gkrumbach07) [reply frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts:25]: it can be just for after, but it is just to prevent the case where you had a test not cleanup properly so this catches that case

File: frontend/src/__tests__/integration/pages/modelServing/ModelServingGlobal.spec.ts (hunk 5 above)
Comment (Gkrumbach07) [reply frontend/src/__tests__/integration/pages/modelServing/ModelServingGlobal.spec.ts:5]: it is going to the storybook page without any of the extra controls around it. this procedure is outlined in the testing doc.

File: frontend/src/__tests__/integration/pages/projects/ProjectDetails.spec.ts (hunk 6 above)
Comment (Gkrumbach07) [reply frontend/src/__tests__/integration/pages/projects/ProjectDetails.spec.ts:5]: what do you suggest? this is how the storybook docs integrate with playwright. https://storybook.js.org/docs/react/writing-tests/importing-stories-in-tests#example-with-playwright

File: package.json (hunk 7 above)
Comment (Gkrumbach07) [reply package.json:57]: ill move it

File: frontend/src/__tests__/integration/pages/projects/ProjectView.spec.ts (hunk 10 above)
Comment (Gkrumbach07) [reply frontend/src/__tests__/integration/pages/projects/ProjectView.spec.ts:38]: I outline this in the testing doc.

Storybook now only provides the mocking and setup for tests. The play function will either do nothing or get to the correct view. (this is for cases where we cant directly navigate to a view like a delete modal)

playwright is now used on top of these stories and implements all the tests. Playwright tests dont show up in the storybook UI, but instead are run by calling `npm run test:integration`

File: frontend/playwright.config.ts (hunk 3 above)
Comment (andrewballantyne) [reply frontend/playwright.config.ts:26]: ah gotcha 👍

File: frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts (hunk 4 above)
Comment (andrewballantyne) [reply frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts:25]: Would you not want to just make the same effort in both spots? It is literal duplication... `await cleanupTestProject(page)` or something...

File: frontend/src/__tests__/integration/pages/modelServing/ModelServingGlobal.spec.ts (hunk 5 above)
Comment (andrewballantyne) [reply frontend/src/__tests__/integration/pages/modelServing/ModelServingGlobal.spec.ts:5]: This needs to be done fair nicer I think... maybe some utility method? I am confused on what exactly is happening in that path -- but it seems boilerplatey and nonsensical to the eye... which screams utility.

File: frontend/src/__tests__/integration/pages/projects/ProjectDetails.spec.ts (hunk 6 above)
Comment (andrewballantyne) [reply frontend/src/__tests__/integration/pages/projects/ProjectDetails.spec.ts:5]: `await goTo(page, 'something', 'other', 'than', 'this')` ... just easier... if there is complexity here to the path generated, maybe: `await goTo(page, { page: 'test', other: 'thing', ...)`

Is this generated by something that is just copy & paste?

File: frontend/src/__tests__/integration/pages/projects/ProjectView.spec.ts (hunk 10 above)
Comment (andrewballantyne) [reply frontend/src/__tests__/integration/pages/projects/ProjectView.spec.ts:38]: Hmmm... okay. I need to read the doc better, I skimmed and this caused me confusion I bet.

File: frontend/package.json [hunk 11]
```diff
@@ -21,23 +21,23 @@
...
-    "test:local": "jest --watchAll",
     "test:type-check": "tsc --noEmit",
-    "test:storybook": "test-storybook --verbose",
-    "test:storybook-ci": "npx concurrently -k -s first -n \"SB,TEST\" -c \"magenta,blue\" \"npm run build:storybook --quiet && npx http-server storybook-static --port 6006 --silent\" \"npx wait-on tcp:127.0.0.1:6006 && npm run test:storybook\"",
-    "storybook": "start-storybook -p 6006 -s ./config"
+    "test:jest": "jest",
+    "test:unit": "npm run test:jest",
+    "test:accessibility": "test-storybook --verbose",
+    "test:integration": "playwright test ./src/__tests__/integration",
+    "test:integration-ci": "npx concurrently -k -s first -n \"SB,TEST\" \"npm run build:storybook --quiet && npx http-server storybook-static --port 6006 --silent\" \"npx wait-on tcp:127.0.0.1:6006 && npm run test:accessibility && npm run test:integration\"",
```
Comment (andrewballantyne) [inline frontend/package.json:38]: Running `npm run test` is now not completable... Why did we bring a CI element into a command line argument for testing?

File: frontend/package.json [hunk 12]
```diff
@@ -21,23 +21,23 @@
...
+    "test:e2e": "export E2E=1; playwright test ./src/__tests__/e2e",
+    "storybook": "storybook dev -p 6006 -s ../config"
```
Comment (andrewballantyne) [inline frontend/package.json:40]: Aside from serving up the infra for other tests, does this have any value anymore?

File: frontend/package.json [hunk 13]
```diff
@@ -21,23 +21,23 @@
...
     "test:type-check": "tsc --noEmit",
-    "test:storybook": "test-storybook --verbose",
-    "test:storybook-ci": "npx concurrently -k -s first -n \"SB,TEST\" -c \"magenta,blue\" \"npm run build:storybook --quiet && npx http-server storybook-static --port 6006 --silent\" \"npx wait-on tcp:127.0.0.1:6006 && npm run test:storybook\"",
-    "storybook": "start-storybook -p 6006 -s ./config"
+    "test:jest": "jest",
+    "test:unit": "npm run test:jest",
+    "test:accessibility": "test-storybook --verbose",
+    "test:integration": "playwright test ./src/__tests__/integration",
+    "test:integration-ci": "npx concurrently -k -s first -n \"SB,TEST\" \"npm run build:storybook --quiet && npx http-server storybook-static --port 6006 --silent\" \"npx wait-on tcp:127.0.0.1:6006 && npm run test:accessibility && npm run test:integration\"",
+    "test:e2e": "export E2E=1; playwright test ./src/__tests__/e2e",
```
Comment (andrewballantyne) [inline frontend/package.json:39]: This does not work for me -- what am I missing?

I have client and server running, I run this and I always get a failure on:
```
 Received: {"_frame": {"_guid": "frame@6f12dfd3c3b8a82344ecf9dba74d88fa", "_type": "Frame"}, "_selector": "internal:text=\"Error creating project\"i"}

      33 |   await page.getByLabel('Name *', { exact: true }).fill('e2e-test-project');
      34 |   await page.getByRole('button', { name: 'Create' }).click();
    > 35 |   await expect(page.getByText('Error creating project')).toBeFalsy();
         |                                                          ^
      36 |   await expect(page.locator('tr', { has: page.locator('text="e2e-test-project"') })).toBeTruthy();
      37 |   await page.getByRole('link', { name: 'Data Science Projects', exact: true }).click();
      38 |

        at /Users/andrewballantyne/code/others/odh-dashboard/frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts:35:58
```

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts
Comment (andrewballantyne) [inline frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts:1]: I wonder if we can make `e2e-test-project` a configurable env var... if you want we can prefix it... but if I want to run the test, and you want to run the test we will collide and I won't know who created what or how if I want to clean it up...

File: frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts (hunk 4 above)
Comment (Gkrumbach07) [reply frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts]: i had there for testing the test purposes. it doesn't need to be in the before

File: frontend/src/__tests__/integration/pages/modelServing/ModelServingGlobal.spec.ts (hunk 5 above)
Comment (Gkrumbach07) [reply frontend/src/__tests__/integration/pages/modelServing/ModelServingGlobal.spec.ts:5]: it kind of is. but you would need to supply the path:

`tests-integration-pages-modelserving-modelservingglobal--delete-model` is coming from `__tests__/integration/pages/modelServing/ModelServingGlobal.tsx` And then the story named `DeleteModel`

in my opinion its more work to pick apart your path for a util function compared to just copy pasting the url from storybook

File: frontend/src/__tests__/integration/pages/projects/ProjectView.spec.ts (hunk 10 above)
Comment (Gkrumbach07) [reply frontend/src/__tests__/integration/pages/projects/ProjectView.spec.ts:38]: i will go over the doc again and remove some fluff to make things more clear

File: frontend/package.json (hunk 12 above)
Comment (Gkrumbach07) [reply frontend/package.json:40]: i think so. it is good for test any a11y errors and making sure all the mocks are working correctly

File: frontend/package.json (hunk 11 above)
Comment (Gkrumbach07) [reply frontend/package.json:38]: what do you mean? does `npm run test` not work for you? should i have the `test` not build storybook and instead have a `test-ci`

File: frontend/package.json (hunk 13 above)
Comment (Gkrumbach07) [reply frontend/package.json:39]: i will look into it

File: frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts
Comment (Gkrumbach07) [reply frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts:1]: will add that. sounds like a good idea

File: frontend/src/__tests__/integration/pages/modelServing/ModelServingGlobal.spec.ts (hunk 5 above)
Comment (andrewballantyne) [reply frontend/src/__tests__/integration/pages/modelServing/ModelServingGlobal.spec.ts:5]: ooooh, I see... interesting. I think we need some utility here based off the path to the file -- but okay, we can leave it for now.

File: frontend/package.json (hunk 11 above)
Comment (andrewballantyne) [reply frontend/package.json:38]: Oh interesting -- maybe because I didn't build storybook? I'll give that a shot... it never completes when I run test, server running or not.

File: frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts
Comment (andrewballantyne) [reply frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts:1]: Is this change implemented? I don't see an env var in your `.env.test` file

File: frontend/package.json (hunk 11 above)
Comment (andrewballantyne) [reply frontend/package.json:38]:
```
[TEST] npx wait-on tcp:127.0.0.1:6006 && npm run test:accessibility && npm run test:integration exited with code 0
--> Sending SIGTERM to other processes..
[SB] npm run build:storybook --quiet && npx http-server storybook-static --port 6006 --silent exited with code SIGTERM
```
@Gkrumbach07 this is what I get when running `npm run test`

File: frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts (hunk 4 above)
Comment (Gkrumbach07) [reply frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts]: i moved it into a helper function and run it before and after now

File: frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts
Comment (Gkrumbach07) [reply frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts:1]: should be? 
```
TEST_PROJECT_NAME=e2e-test-project
```

File: frontend/package.json (hunk 13 above)
Comment (Gkrumbach07) [reply frontend/package.json:39]: e2e has been fixed. there was issues with the expect calls

File: frontend/package.json (hunk 11 above)
Comment (Gkrumbach07) [reply frontend/package.json:38]: it should be running a bunch of other tests too. those are the only lines its putting out?

File: frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts [hunk 14]
```diff
@@ -0,0 +1,69 @@
+import { test, expect, Page } from '@playwright/test';
+
+const cleanTestProject = async (page: Page) => {
+  await page.goto('/projects');
+  const count = await page.getByRole('link', { name: process.env.TEST_PROJECT_NAME }).count();
+  if (count > 0) {
+    await page
+      .locator('tr', {
+        has: page.locator(`text="${process.env.TEST_PROJECT_NAME}"`),
```
Comment (andrewballantyne) [inline frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts:9]: Let us not refer to the process var directly -- please put it in the constants files

Comment (andrewballantyne) [reply frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts:9]: It seems we have a horrible organization on this front too 🤦 More tech debt. 

Today the constants are in here: `frontend/src/utilities/const.ts` -- it's awkward, but please put it in there too.

File: frontend/package.json (hunk 11 above)
Comment (andrewballantyne) [reply frontend/package.json:38]: Those are the last ones 😞 -- before that is it executing the previous test

File: frontend/src/__tests__/integration/pages/projects/ProjectView.spec.ts (hunk 10 above)
Comment (Gkrumbach07) [reply frontend/src/__tests__/integration/pages/projects/ProjectView.spec.ts:38]: updated the doc. should be more of a step by step guide now

File: frontend/package.json (hunk 11 above)
Comment (Gkrumbach07) [reply frontend/package.json:38]: i think i resolved the issue. i was searching for a button with the text "Create", however this was searching for a non exact match so it was finding other buttons with the word create in it. I am not sure why it was passing for me locally and in CI for some of the runs

File: frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts (hunk 14 above)
Comment (Gkrumbach07) [reply frontend/src/__tests__/e2e/CreateEditDeleteProject.spec.ts]: Issues with playwright and typescript path mapping

**Review by andrewballantyne — APPROVED** [review]
